src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs

```diff
@@ -342,28 +283,13 @@ void IConnectionControl.End(ProduceEndType endType)
             switch (endType)
             {
                 case ProduceEndType.ConnectionKeepAlive:
-                    if (_connectionState != ConnectionState.Open)
-                    {
-                        return;
-                    }
-
                     Log.ConnectionKeepAlive(ConnectionId);
                     break;
                 case ProduceEndType.SocketShutdown:
                 case ProduceEndType.SocketDisconnect:
-                    lock (_stateLock)
-                    {
-                        if (_connectionState == ConnectionState.Disconnecting ||
-                            _connectionState == ConnectionState.SocketClosed)
-                        {
-                            return;
-                        }
-                        _connectionState = ConnectionState.Disconnecting;
-
-                        Log.ConnectionDisconnect(ConnectionId);
-                        _rawSocketOutput.End(endType);
-                        break;
-                    }
+                    Log.ConnectionDisconnect(ConnectionId);
```

Comment:
Might be better to log this in `OnSocketClosed` now. It was arguably the better place to begin with, and we shouldn't risk duplicate logs when it's so easily avoidable.